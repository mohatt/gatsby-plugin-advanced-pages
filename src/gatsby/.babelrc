{
  "presets": [
    ["babel-preset-gatsby-package", { "nodeVersion": "8" }]
  ]
}

