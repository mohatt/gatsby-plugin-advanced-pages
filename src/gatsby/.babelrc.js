module.exports = {
  presets: [
    ['babel-preset-gatsby-package', {
      browser: false,
      nodeVersion: '18.0.0',
      esm: !!process.env.BUILD_ESM,
    }]
  ]
}
