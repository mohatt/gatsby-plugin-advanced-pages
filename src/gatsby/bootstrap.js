import { initializeOptions, initializeReporter } from './util'
import { optionsSchema } from './schema'

// Validates user-defined options against schema
// runs before onPreBootstrap
export function pluginOptionsSchema ({ Joi }) {
  return optionsSchema(Joi)
}

// Initializes plugin reporter
export function onPreInit ({ reporter }) {
  initializeReporter(reporter)
}

// Initializes plugin options
export function onPreBootstrap ({ store, reporter }, pluginOptions) {
  // Default values for options
  const defaultOptions = {
    basePath: '/',
    pages: [],
    template: null,
    directories: {
      templates: './src/templates',
      helpers: './gatsby/pages'
    },
    pagination: {
      limit: 10,
      suffix: '/page/:page'
    },
    typeNames: {
      page: 'Page'
    }
  }

  // Initializes and validates options
  // Only runs once at bootstrap
  initializeOptions({
    store,
    defaultOptions,
    pluginOptions
  })
}
