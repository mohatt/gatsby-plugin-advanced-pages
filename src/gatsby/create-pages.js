import path from 'path'
import PagesCreator from './lib/pages-creator'
import { getOption, reportError } from './util'

export default async function ({ graphql, actions, getNodesByType }) {
  const { createPage } = actions

  const pageType = getOption('typeNames.page')
  if(getNodesByType(pageType).length === 0) {
    return
  }

  const result = await graphql(`
    {
      all${pageType} {
        nodes {
          id
          routes {
            name
            path
          }
          title
          templateName
          template
          helper
        }
      }
    }
  `)

  if (result.errors) {
    return reportError('Failed running "create-pages" GraphQL Query', result.errors.shift())
  }

  const pageCreator = new PagesCreator(
    result.data[`all${pageType}`].nodes
  )

  // Create the actual pages
  await pageCreator.createPages({ graphql, createPage })

  // Write routes export file
  pageCreator.writeRoutesExport(path.resolve(__dirname, '../routes.js'))
}
