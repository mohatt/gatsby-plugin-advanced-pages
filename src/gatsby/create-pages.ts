import fs from 'fs'
import path from 'path'
import type { GatsbyNode } from 'gatsby'
import PagesCreator from './lib/pages-creator'
import { options, reporter } from './util'

const createPages: GatsbyNode['createPages'] = async ({ graphql, actions, getNodesByType, cache }) => {
  const { createPage } = actions

  const pageType = options.get('typeNames').page
  if (getNodesByType(pageType).length === 0) {
    return
  }

  const result = await graphql(`
    {
      all${pageType} {
        nodes {
          id
          routes {
            name
            path
          }
          title
          templateName
          template
          helper
        }
      }
    }
  `)

  if (result.errors) {
    reporter.error('Failed running "create-pages" GraphQL Query', result.errors.shift())
    return
  }

  // Create the actual pages
  const pageCreator = new PagesCreator(result.data[`all${pageType}`].nodes)
  await pageCreator.createPages({ graphql, createPage })

  // Write routes export file
  const routesFile = path.join(cache.directory, 'routes.json')
  try {
    fs.writeFileSync(routesFile, JSON.stringify(pageCreator.getRoutesExport()))
  } catch (e) {
    reporter.error('Error writing route map export file', e)
  }
}

export default createPages
