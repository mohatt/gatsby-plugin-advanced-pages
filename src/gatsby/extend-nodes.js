const { getOption } = require('./util')

// Extend frontmatter type definition
module.exports = ({ type }) => {
  if (type.name === getOption('engine.typeName')) {
    return {
      'frontmatter.title': 'String',
      'frontmatter.routes': 'JSON',
      'frontmatter.template': 'String',
      'frontmatter.helper': 'String',
      'frontmatter.data': 'JSON',
    }
  }

  return {}
}
