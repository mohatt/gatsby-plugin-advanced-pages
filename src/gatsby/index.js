export { default as onPreBootstrap } from './bootstrap'
export { default as sourceNodes } from './source-nodes'
export { default as setFieldsOnGraphQLNodeType } from './extend-nodes'
export { default as onCreateNode } from './on-create-node'
export { default as createPages } from './create-pages'
export { default as onPreExtractQueries } from './install-fragments'
