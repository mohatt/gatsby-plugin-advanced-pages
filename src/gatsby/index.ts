export type * from './types'
export type { RouteParams, PathGenerator } from '../lib/route-compiler'
