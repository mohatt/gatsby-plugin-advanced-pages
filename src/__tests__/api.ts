import { Mock } from 'vitest'
import { withPrefix, navigate as gatsbyNavigate } from 'gatsby'
import { useLocation } from '@reach/router'
import {
  getRoutes,
  routeExists,
  getRoute,
  useRoute,
  useIsRoute,
  getMatchingRoute,
  generatePath,
  navigate,
} from '../index'

describe('API', () => {
  describe('getRoutes', () => {
    it('correctly fetches routes', () => {
      expect(getRoutes()).toMatchSnapshot()
      expect(getRoutes('blog')).toMatchSnapshot()
    })
  })

  describe('getRoute', () => {
    it('correctly fetches routes', () => {
      expect(getRoute('blog.post')).toMatchSnapshot()
      expect(() => getRoute('invalid')).toThrow()
      expect(() => getRoute(<any>false)).toThrow()
      expect(() => getRoute(<any>23)).toThrow()
    })
  })

  describe('routeExists', () => {
    it('correctly checks for valid routes', () => {
      expect(routeExists('blog.post')).toBe(true)
      expect(routeExists('invalid')).toBe(false)
    })
  })

  describe('useRoute', () => {
    it('correctly gets the current active route', () => {
      expect(useRoute()).toMatchSnapshot()
    })

    it('returns null when no route is active', () => {
      ;(<Mock>useLocation).mockReturnValueOnce({ pathname: '/' })
      expect(useRoute()).toBe(null)
    })
  })

  describe('useIsRoute', () => {
    it('picks the correct matching route', () => {
      expect(useIsRoute('home')).toBe(false)
      expect(useIsRoute('blog')).toBe(true)
      expect(() => useIsRoute('invalid')).toThrow()
    })

    it('returns false when no route is active', () => {
      ;(<Mock>useLocation).mockReturnValueOnce({ pathname: '/' })
      expect(useIsRoute('blog')).toBe(false)
    })
  })

  describe('getMatchingRoute', () => {
    it('picks the correct matching route', () => {
      expect(getMatchingRoute('/')).toMatchSnapshot()
      expect(getMatchingRoute(withPrefix('/'), true)).toMatchSnapshot()
      expect(getMatchingRoute('/blog/tag/test/page/4')).toMatchSnapshot()
      expect(getMatchingRoute('/blog/post/hi')).toMatchSnapshot()
    })
  })

  describe('generatePath', () => {
    it('correctly generates paths', () => {
      expect(generatePath('home')).toMatchSnapshot()
      expect(generatePath('blog', { page: 5 }, 'pagination')).toMatchSnapshot()
      expect(generatePath('blog.post', { post: 'hello' })).toMatchSnapshot()
      expect(() => generatePath('invalid')).toThrow()
    })

    it('throws for unrecognized scope', () => {
      expect(() => generatePath('about', {}, 'pagination')).toThrowErrorMatchingSnapshot()
    })
  })

  describe('navigate', () => {
    it('correctly calls gatsby.navigate()', async () => {
      const gMock = <Mock>gatsbyNavigate
      await navigate('home')
      expect(gMock.mock.calls).toMatchSnapshot()
      gMock.mockClear()
      await navigate('blog.post', { post: 'hello' })
      expect(gMock.mock.calls).toMatchSnapshot()
      gMock.mockClear()
      await navigate('blog', { page: 4 }, 'pagination', { replace: true })
      expect(gMock.mock.calls).toMatchSnapshot()
      gMock.mockClear()
    })
  })
})
