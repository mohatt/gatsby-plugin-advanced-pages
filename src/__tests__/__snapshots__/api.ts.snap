// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`API > generatePath > correctly generates paths 1`] = `"/site"`;

exports[`API > generatePath > correctly generates paths 2`] = `"/site/blog/page/5"`;

exports[`API > generatePath > correctly generates paths 3`] = `"/site/blog/post/hello"`;

exports[`API > generatePath > throws for unrecognized scope 1`] = `[TypeError: Unrecognized scope 'pagination' on route 'about']`;

exports[`API > getMatchingRoute > picks the correct matching route 1`] = `
{
  "name": "home",
  "parent": null,
  "path": "/",
  "realpath": "/site/",
}
`;

exports[`API > getMatchingRoute > picks the correct matching route 2`] = `
{
  "name": "home",
  "parent": null,
  "path": "/",
  "realpath": "/site/",
}
`;

exports[`API > getMatchingRoute > picks the correct matching route 3`] = `
{
  "name": "blog.tag.pagination",
  "parent": {
    "name": "blog.tag",
    "scope": "pagination",
  },
  "path": "/blog/tag/:tag/page/:page",
  "realpath": "/site/blog/tag/:tag/page/:page",
}
`;

exports[`API > getMatchingRoute > picks the correct matching route 4`] = `
{
  "name": "blog.post",
  "parent": null,
  "path": "/blog/post/:post",
  "realpath": "/site/blog/post/:post",
}
`;

exports[`API > getRoute > correctly fetches routes 1`] = `
{
  "name": "blog.post",
  "parent": null,
  "path": "/blog/post/:post",
  "realpath": "/site/blog/post/:post",
}
`;

exports[`API > getRoutes > correctly fetches routes 1`] = `
[
  {
    "name": "home",
    "parent": null,
    "path": "/",
    "realpath": "/site/",
  },
  {
    "name": "about",
    "parent": null,
    "path": "/about",
    "realpath": "/site/about",
  },
  {
    "name": "blog",
    "parent": null,
    "path": "/blog",
    "realpath": "/site/blog",
  },
  {
    "name": "blog.pagination",
    "parent": {
      "name": "blog",
      "scope": "pagination",
    },
    "path": "/blog/page/:page",
    "realpath": "/site/blog/page/:page",
  },
  {
    "name": "blog.author",
    "parent": null,
    "path": "/blog/author/:author",
    "realpath": "/site/blog/author/:author",
  },
  {
    "name": "blog.author.pagination",
    "parent": {
      "name": "blog.author",
      "scope": "pagination",
    },
    "path": "/blog/author/:author/page/:page",
    "realpath": "/site/blog/author/:author/page/:page",
  },
  {
    "name": "blog.tag",
    "parent": null,
    "path": "/blog/tag/:tag",
    "realpath": "/site/blog/tag/:tag",
  },
  {
    "name": "blog.tag.pagination",
    "parent": {
      "name": "blog.tag",
      "scope": "pagination",
    },
    "path": "/blog/tag/:tag/page/:page",
    "realpath": "/site/blog/tag/:tag/page/:page",
  },
  {
    "name": "blog.post",
    "parent": null,
    "path": "/blog/post/:post",
    "realpath": "/site/blog/post/:post",
  },
]
`;

exports[`API > getRoutes > correctly fetches routes 2`] = `
[
  {
    "name": "blog.pagination",
    "parent": {
      "name": "blog",
      "scope": "pagination",
    },
    "path": "/blog/page/:page",
    "realpath": "/site/blog/page/:page",
  },
]
`;

exports[`API > navigate > correctly calls gatsby.navigate() 1`] = `
[
  [
    "/site",
    undefined,
  ],
]
`;

exports[`API > navigate > correctly calls gatsby.navigate() 2`] = `
[
  [
    "/site/blog/post/hello",
    undefined,
  ],
]
`;

exports[`API > navigate > correctly calls gatsby.navigate() 3`] = `
[
  [
    "/site/blog/page/4",
    {
      "replace": true,
    },
  ],
]
`;

exports[`API > useRoute > correctly gets the current active route 1`] = `
{
  "name": "blog.pagination",
  "parent": {
    "name": "blog",
    "scope": "pagination",
  },
  "path": "/blog/page/:page",
  "realpath": "/site/blog/page/:page",
}
`;
